File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.cc
#include <bits/stdc++.h>
using namespace std;
int n, s, nn;
vector<int> v;
int cnt;

void dfs(int idx, int sum) {
	if(idx == n) {
		if(sum == s) {
			cnt++;
		}
		return;
	}
	dfs(idx+1, sum+v[idx]);
	dfs(idx+1, sum);
}

int main(int argc, char** argv)
{   
	cin >> n >> s;
	for(int i=0; i<n; i++) {
		cin >> nn;
		v.push_back(nn);
	}
	cnt = 0;
	dfs(0, 0);
	if(s == 0) cnt--;
	cout << cnt;
	return 0;
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
    static boolean[] vis;
    static int[] arr;
    static int n, s;
    static int cnt;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        s = Integer.parseInt(st.nextToken());

        vis = new boolean[n];
        arr = new int[n];
        cnt = 0;
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        dfs(arr, 0, 0, 0);
        System.out.println(cnt);
    }

    static void dfs(int[] arr, int sum, int depth, int idx) {
        //System.out.println("depth: " + depth + " sum: " + sum + " s: " + s);
        if(idx == n) {
            if(depth != 0 && sum == s) cnt++;
            return;
        }
        dfs(arr, sum+arr[idx], depth+1, idx+1);
        dfs(arr, sum, depth, idx+1);
    }
}